src/Catalyst.Core.Modules.Dfs/WebApi/V0/Controllers/DhtController.cs

```diff
@@ -0,0 +1,148 @@
+
+        /// <summary>
+        ///   The peer that has the information.
+        /// </summary>
+        internal IEnumerable<DhtPeerResponseDto> Responses;
+
+        /// <summary>
+        ///   Unknown.
+        /// </summary>
+        public string Extra = string.Empty;
+    }
+
+    /// <summary>
+    ///   Information on a peer that has the information.
+    /// </summary>
+    internal sealed class DhtPeerResponseDto
+    {
+        /// <summary>
+        ///   The peer ID.
+        /// </summary>
+        public string Id;
+
+        /// <summary>
+        ///   The listening addresses of the peer.
+        /// </summary>
+        public IEnumerable<string> Addrs;
+    }
+
+    /// <summary>
+    ///   Distributed Hash Table.
+    /// </summary>
+    /// <remarks>
+    ///   The DHT is a place to store, not the value, but pointers to peers who have 
+    ///   the actual value.
+    /// </remarks>
+    public sealed class DhtController : IpfsController
+    {
+        /// <summary>
+        ///   Creates a new controller.
+        /// </summary>
+        public DhtController(ICoreApi dfs) : base(dfs) { }
+
+        /// <summary>
+        ///   Query the DHT for all of the multiaddresses associated with a Peer ID.
+        /// </summary>
+        /// <param name="arg">
+        ///   The peer ID to find.
+        /// </param>
+        /// <returns>
+        ///   Information about the peer.
+        /// </returns>
+        [HttpGet, HttpPost, Route("dht/findpeer")]
+        public async Task<DhtPeerDto> FindPeer(string arg)
+        {
+            var peer = await IpfsCore.DhtApi.FindPeerAsync(arg, Cancel);
+            return new DhtPeerDto
+            {
+                Id = peer.Id.ToBase58(),
+                Responses = new[]
+                {
+                    new DhtPeerResponseDto
+                    {
+                        Id = peer.Id.ToBase58(),
+                        Addrs = peer.Addresses.Select(a => a.WithoutPeerId().ToString())
+                    }
+                }
+            };
+        }
+
+        /// <summary>
+        ///  Find peers in the DHT that can provide a specific value, given a key.
+        /// </summary>
+        /// <param name="arg">
+        ///   The CID key,
+        /// </param>
+        /// <param name="limit">
+        ///   The maximum number of providers to find.
+        /// </param>
+        /// <returns>
+        ///   Information about the peer providers.
+        /// </returns>
+        [HttpGet, HttpPost, Route("dht/findprovs")]
+        public async Task<IEnumerable<DhtPeerDto>> FindProviders(string arg,
+            [ModelBinder(Name = "num-providers")] int limit = 20)
+        {
+            var peers = await IpfsCore.DhtApi.FindProvidersAsync(arg, limit, null, Cancel);
+            return peers.Select(peer => new DhtPeerDto
+            {
+                Id = peer.Id.ToBase58(), // TODO: should be the peer ID that answered the query
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Complete the task associated to this 'TODO' comment.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)